tool/teleport/common/sftp.go

```diff
@@ -75,17 +79,96 @@ func (c compositeCh) Close() error {
 	return trace.NewAggregate(c.r.Close(), c.w.Close())
 }
 
+type byteReader struct {
+	buf []byte
+	io.Reader
+}
+
+func newByteReader(r io.Reader) *byteReader {
+	return &byteReader{
+		Reader: r,
+		buf:    make([]byte, 1),
+	}
+}
+
+func (r *byteReader) ReadByte() (byte, error) {
+	_, err := r.Reader.Read(r.buf)
+	if err != nil {
+		return 0, err
+	}
+
+	return r.buf[0], nil
```

Comment:
This looks very dangerous to me. If you are reading from a source that has a value of `[]byte{1, 2, 3, 4, 5}`, and the first read returns `[]byte{1, 2}`, you'll silently drop the `2`.

Reply:
Unless I'm missing something, `byteReader.buf` always has a size of 1 so a read should never return more than 1 byte. I'll add a check just in case though.